### ISSUE TITLE
Web UI Bugs: reset doesn't reset + other cosmetic glitches

### ISSUE THREAD

**Issue by robertlipe** [PlummersSoftwareLLC/NightDriverStrip#411]
## Bug report
The reset/restart button in the the web UI doesn't restart remaining count.

**Problem**
Prev and Next work as expected. I'd expec the 'reset' button to reset the remaining time on this effect. It just doesn't.

**Steps**
1.  Click <home> (Why is this step needed? Why isn't 'home' 'home'?
2.  Click on next and prev a few times. Watch current effect cycle. This verifies that your clicking finger works. :-)
3.  Click on reset. Nothing visible changes. I'd expect counter to reset to the cumber on the left.

**Example**
It's not intermittent. It just doesn't seem to be hooked up.

Bonus points to the assignee of this bug for reconsidering if milliseconds really is the appropriate unit of measure to show (without units or a decimal place) and to accept for input on 'interval'. If you type '120' for interval, is it really because you wanted 8 changes per second?

Bonus Bonus bug (maybe this should be three reports...) if you set the interval to '1000', the 'time remaining' field flashes for a cycle or two, but decided it can't keep up and just displays a constant number. Seems it should blank or something. Or maybe we don't allow setting times below a second.

Bonus Bonus Bonue (OK, four reports) if you type a negative number (!) it sets the time to 2^32 - 10000. Time remaining then overlaps the "Previojs" button.

**Notes**
<img width="437" alt="image" src="https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/8724975/bc2b2829-df38-4ddf-8a38-49774ac7e485">

<img width="321" alt="image" src="https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/8724975/8b96fbd4-5d9e-492d-9e76-0f75ab87bcc5">

I aborted nothing. What are the other five errors?

**Comment by rbergen** [comment]
TL;DR regarding the main report: the "reset" button behavior is the expected behavior.

The reset button is actually a "refresh" button. The interaction between the UI and the API is currently pull-only, which means that the interval timer updates _and the tracking of what effect is playing_ are largely done in the browser. The actual state is pulled from the device API at some meaningful interval, i.e. seconds. 
Specifically when the web UI browser tab is not shown, this means the device and the UI can/will get out of sync. The refresh button forces the UI to pull the current state of affairs from the device API, to resync what the UI is showing. 

Issue #302 includes a few comments about the merits of adding (a) web socket(s) to the on-device web server implementation; one could be used to "push" on-device effect changes to the web UI. On top of that, issue #356 describes a feature that would absolutely require one.

As for the other points: I think setting the timer in milliseconds is indeed way too granular. Showing the countdown in some fraction of seconds may make sense, but maybe tens of seconds would be enough.
Setting the timeout in seconds would automatically address bug 3.

Bug 4 I'd have to look into, which I'm currently not able to. I'll try to remember to do so at a later time.

**Comment by KeiranHines** [comment]
I am late to the party here @robertlipe Bug 1 and Bug 4 should be fixed in the main branch now. As of #431

**Comment by rbergen** [comment]
@KeiranHines Just for my understanding: what's Bug 1 in your reference?

**Comment by KeiranHines** [comment]
Ahh yes that wasn't clear at all. 
Bug 1
> Click (Why is this step needed? Why isn't 'home' 'home'?

and Bug 4

> Bonus Bonus Bonue (OK, four reports) if you type a negative number (!) it sets the time to 2^32 - 10000. Time remaining then overlaps the "Previojs" button.

have been fixed.

**Comment by rbergen** [comment]
Check, thanks for clarifying that!

**Comment by rbergen** [comment]
The remaining time for an effect is determined by the back-end; the front-end should just report what the back-end is doing. When the effect changes, the "remaining time" at the back-end will always reset. If this is not reflected in the front-end, then that is a bug.

**Comment by rbergen** [comment]
No, UpdateOTA does not force a sync of anything. A reboot triggered via the /reset endpoint does.

**Comment by rbergen** [comment]
From the device's firmware perspective (i.e. what we control) the upload via serial is not manageable in any way. We don't get told and we don't get the chance to say "hang on a second while we save this". The upload is triggered at a lower level and when it starts the firmware is aborted wherever it is at the time.

Concerning persistence, when it comes to this part of it, it's simpler than you seem to think. When the running effect changes, all that's persisted after a few seconds is the index of the newly started effect in the effect list. This means that if the device is rebooted after an effect change (and the new effect's index has a chance to be persisted) then the effect that was running at the time of reboot will be started again, and it will run the full interval.

Pinning the running effect is already possible. If you set the default interval time to 0, EffectManager will actually never switch effects. I'm not totally sure if the Web UI will let you do this, but I know the ON button on the remote actually does exactly that. That is to say, pressing ON on the remote will freeze the effect cycle at whichever is the current effect. So, if you set the effect time to 0 (like by pressing ON on the remote) and then switch to the effect you want it to stay at, it will. Due to the effect index persistence I just described and the fact that the effect interval is also persisted, this behavior will persist across reboots as well until you change the effect interval to a value other than 0.

**Comment by rbergen** [comment]
> If you have a web session open and are watching the 120 qadzillion
zilliseconds march down and beat on Brightness Up or Brightness Down (sigh)
to change effect, I have confirmed that the timer does NOT get reset to
12000000000000; it continues marching on. Its' like there's a timer in the
web app and a timer on the board and they're not REALLY synchronized.

Ah, you're changing the effect using the IR remote, and flagging that the "time remaining" indicator in the web UI doesn't change along when you do. In itself, that is a new bug in my opinion. What I'm curious about is if the web UI _does_ represent that the effect has changed when this happens?  
If not, we may "just" be looking at the fact that the device does not "push" changes that happen to the web UI that _may_ be relevant for what the web UI is showing. One reason for that is that there currently is no channel in place that allows this to happen. This would require a web socket to be added to the on-board webserver mix, which currently just isn't there.

**Comment by KeiranHines** [comment]
That working theory seems pretty valid. By memory the useEffect hook managing the time update. I'd have to take a look at it closer to understand it's update mechanics.

**Comment by rbergen** [comment]
> Though the web was the boss that told the board to change, the web
didn't reset its own timer.

What's really important here is if the effect change was effectuated through the IR remote or not. If so, it makes sense the web UI is unaware.

**Comment by rbergen** [comment]
> That working theory seems pretty valid. By memory the useEffect hook managing the time update. I'd have to take a look at it closer to understand it's update mechanics.

I know for certain that the web front-end keeps its own timer. It does not retrieve the amount of time left from the device every microsecond - for one because it technically can't. My point is basically that:
- if changes are made via the web UI (like changing an effect, or the effect interval, or ...) that don't reflect in the web UI itself, then that's a bug in the web UI. 
- if similar changes are made via another channel (like the infrared remote control) then the web UI currently can't know that they happen exactly when they happen. That's not so much a bug as a limitation in our current software architecture - that being that updates that take place on the device can't be pushed to (potentially) interested parties outside of it.

**Comment by KeiranHines** [comment]
Some work has been done to improve this in #477 There is now a global effect state and a period refresh call to the `/effects` endpoint (currently 30s) to try to improve the sync between what the UI has and what the device has.

**Comment by rbergen** [comment]
@robertlipe Just for your information: a few of the things you mention will be addressed by #477. Some others won't (yet).

**Comment by rbergen** [comment]
With the merge of #532, I believe the points in this issue that matter have been addressed. It would still be nice if the front-end picks up (effect) changes that happen at the back-end, but that is now covered by the discussion in #356.

@robertlipe Are you okay with me closing this?
